### ISSUE TITLE
Run 2 - Kitchen Sink

### ISSUE THREAD

**Issue by MatthewLacerda2** [MatthewLacerda2/TinyRefinementModel#5]
We just had our first training night. Learned a lot:

- [x] Hyperparameters suck
- [x] FSDP - Than double latent dim and halve the batch size
- [x] Use exponential loss for the diversity of the slots
- [x] Latent Semantic Loss
- [x] RLVR
- [x] Improve Dataset
- [x] Halting head is broken

**Comment by MatthewLacerda2** [comment]
The model iterates on a continuous latent state (expected_shared). We can calculate the distance between the model's final latent thought vector and the embedding vector of the ground-truth answer.

Add a Mean Squared Error (MSE) loss directly to the train_step:$$L_{MSE} = || \text{expected\_shared}_{final} - \text{embed}(y_{target}) ||^2$$

The Catch: "Representation Collapse." If your sole loss function is to make your predicted embedding match the target embedding, neural networks are notoriously lazy. The model might discover an adversarial "junk" vector that mathematically minimizes the $L_{MSE}$ distance but decodes into absolute gibberish when passed to your final vocabulary layer. You have to carefully balance this embedding distance loss with a standard token-prediction loss to ensure the latent space remains tied to human vocabulary.

**Comment by MatthewLacerda2** [comment]
Latent Semantic Loss kinda replaced RLVR. We compare the output's semantic against the expected output's in order to determine the loss. Ideally you'd compare the correctness of the answer, but this'll do

**Comment by MatthewLacerda2** [comment]
I chose to just train on fineweb, finemath and find some python thing. That'll be it. RLVR and other stuff will be for later

**Comment by MatthewLacerda2** [comment]
- The Latent Semantic Loss was dropped for regular next-token, since its more "accurate" (less forgiving). It was cool tho. Perhaps a RL stack is just better, but we are resource constrained
- The diversity got back at a very regular orthogonal loss, which works very well
- Hyperparameters were a bitch to get right, specially in ACT, no easy task
- Dataset doesn't need to be complex, just use a nice Curriculum and SFT and you're golden
